File: harness/src/fs.rs
//! Harness-side filesystem adapters wrapping the consolidated `shell` worker.
//!
//! The shell worker's `shell::fs::read` returns a streaming `StreamChannelRef`;
//! the harness web FilesystemPanel preview wants inline bytes. This module
//! provides `read_inline` to bridge the two.

use serde::{Deserialize, Serialize};
use serde_json::{json, Value};

/// Default cap for inline reads. Matches the historical
/// `shell-filesystem::read` cap so existing UI behaviour is preserved.
pub const DEFAULT_MAX_INLINE_BYTES: usize = 256 * 1024;

#[derive(Debug, Deserialize)]
pub struct ReadInlineArgs {
    pub path: String,
    #[serde(default)]
    pub max_bytes: Option<usize>,
}

/// Wire-compatible mirror of `shell::fs::ReadResponseWire` so we can
/// deserialize the shell worker's response without depending on the
/// `iii-shell` crate.
#[derive(Debug, Deserialize)]
pub struct ShellFsReadResponse {
    pub content: ContentRef,
    pub size: u64,
    #[allow(dead_code)]
    pub mode: String,
    #[allow(dead_code)]
    pub mtime: i64,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct ContentRef {
    pub channel_id: String,
    pub access_key: String,
    #[serde(default)]
    pub direction: String,
}

/// Pure helper: build the legacy `shell::filesystem::read` envelope from
/// drained bytes plus the upstream `size`. Truncation is detected when
/// `bytes.len() < total_size`.
pub fn build_inline_envelope(bytes: &[u8], total_size: u64) -> Value {
    let bytes_read = bytes.len();
    let truncated = (bytes_read as u64) < total_size;
    let text = std::str::from_utf8(bytes)
        .map_or_else(|_| format!("<binary {} bytes>", bytes_read), str::to_string);
    json!({
        "content": [{ "type": "text", "text": text }],
        "details": {
            "size": total_size,
            "truncated": truncated,
            "bytes_read": bytes_read,
        },
        "terminate": false,
    })
}

use iii_sdk::channels::{ChannelDirection, ChannelReader, StreamChannelRef};
use iii_sdk::{IIIError, TriggerRequest, III};

const READ_TRIGGER_TIMEOUT_MS: u64 = 30_000;

/// Drive `shell::fs::read` and inline up to `max_bytes` of the streamed
/// channel into a legacy `{content:[{text}], details:{...}}` envelope.
///
/// `engine_ws_base` is the same URL passed to `register_worker` in
/// `main.rs`; the harness already threads it through
/// `register_with_iii_with_engine_url`, so callers can pass it directly.
pub async fn read_inline(
    iii: &III,
    engine_ws_base: &str,
    args: ReadInlineArgs,
) -> Result<Value, IIIError> {
    let max_bytes = args.max_bytes.unwrap_or(DEFAULT_MAX_INLINE_BYTES);

    // 1. Ask the shell worker to set up the read channel.
    let raw = iii
        .trigger(TriggerRequest {
            function_id: "shell::fs::read".into(),
            payload: json!({ "path": args.path }),
            action: None,
            timeout_ms: Some(READ_TRIGGER_TIMEOUT_MS),
        })
        .await
        .map_err(|e| IIIError::Handler(format!("shell::fs::read trigger failed: {e}")))?;

    let parsed: ShellFsReadResponse = serde_json::from_value(raw.clone()).map_err(|e| {
        IIIError::Handler(format!(
            "shell::fs::read returned unexpected shape: {e} (raw: {raw})"
        ))
    })?;

    // 2. Drain the channel into a capped buffer.
    let chan_ref = StreamChannelRef {
        channel_id: parsed.content.channel_id.clone(),
        access_key: parsed.content.access_key.clone(),
        direction: ChannelDirection::Read,
    };
    let reader = ChannelReader::new(engine_ws_base, &chan_ref);
    let mut buf: Vec<u8> = Vec::new();
    while buf.len() < max_bytes {
        match reader.next_binary().await {
            Ok(Some(chunk)) => {
                let remaining = max_bytes.saturating_sub(buf.len());
                if chunk.len() <= remaining {
                    buf.extend_from_slice(&chunk);
                } else {
                    buf.extend_from_slice(&chunk[..remaining]);
                    break;
                }
            }
            Ok(None) => break,
            Err(e) => {
                let _ = reader.close().await;
                return Err(IIIError::Handler(format!("channel drain failed: {e}")));
            }
        }
    }
    let _ = reader.close().await;

    // 3. Build the inline envelope (pure, unit-tested in Task 8).
    Ok(build_inline_envelope(&buf, parsed.size))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn build_inline_envelope_round_trips_small_utf8() {
        let env = build_inline_envelope(b"hello", 5);
        assert_eq!(env["content"][0]["type"], "text");
        assert_eq!(env["content"][0]["text"], "hello");
        assert_eq!(env["details"]["size"], 5);
        assert_eq!(env["details"]["truncated"], false);
        assert_eq!(env["details"]["bytes_read"], 5);
        assert_eq!(env["terminate"], false);
    }

    #[test]
    fn build_inline_envelope_marks_truncation_when_drained_is_short() {
        // total_size = 1000 but we only got 256 bytes back.
        let bytes = vec![b'x'; 256];
        let env = build_inline_envelope(&bytes, 1000);
        assert_eq!(env["details"]["size"], 1000);
        assert_eq!(env["details"]["truncated"], true);
        assert_eq!(env["details"]["bytes_read"], 256);
        assert_eq!(env["content"][0]["text"].as_str().unwrap().len(), 256);
    }

    #[test]
    fn build_inline_envelope_emits_binary_marker_for_invalid_utf8() {
        // 0xFF is invalid UTF-8 in any leading position.
        let bytes = vec![0xFF, 0xFE, 0xFD];
        let env = build_inline_envelope(&bytes, 3);
        assert_eq!(env["content"][0]["text"], "<binary 3 bytes>");
        assert_eq!(env["details"]["size"], 3);
        assert_eq!(env["details"]["truncated"], false);
        assert_eq!(env["details"]["bytes_read"], 3);
    }
}

File: harness-node/src/harness/fs.ts
/**
 * `harness::fs::read_inline` — wrap `shell::fs::read` and inline the
 * binary stream into the legacy `{content, details}` envelope. Mirrors
 * `harness/src/fs.rs`.
 */

import { unwrapBody } from '../runtime/handler.js';
import type { ISdk, StreamChannelRef } from '../runtime/iii.js';
import { logger } from '../runtime/otel.js';

export const DEFAULT_MAX_INLINE_BYTES = 256 * 1024;

type ContentRef = { channel_id: string; access_key: string; direction?: string };
type ShellFsReadResponse = {
  content: ContentRef;
  size: number;
  mode?: string;
  mtime?: number;
};

export function buildInlineEnvelope(bytes: Uint8Array, totalSize: number): Record<string, unknown> {
  const bytes_read = bytes.length;
  const truncated = bytes_read < totalSize;
  let text: string;
  try {
    text = new TextDecoder('utf-8', { fatal: true }).decode(bytes);
  } catch {
    text = `<binary ${bytes_read} bytes>`;
  }
  return {
    content: [{ type: 'text', text }],
    details: { size: totalSize, truncated, bytes_read },
    terminate: false,
  };
}

export function register(iii: ISdk, engineWsBase: string): void {
  iii.registerFunction(
    'harness::fs::read_inline',
    async (raw: unknown) => {
      const body = unwrapBody(raw);
      const path = typeof body.path === 'string' ? body.path : null;
      if (!path) throw new Error('missing required field: path');
      const max_bytes =
        typeof body.max_bytes === 'number' && body.max_bytes > 0
          ? body.max_bytes
          : DEFAULT_MAX_INLINE_BYTES;

      const raw2 = await iii.trigger<unknown, ShellFsReadResponse>({
        function_id: 'shell::fs::read',
        payload: { path },
        timeoutMs: 30_000,
      });
      if (!raw2 || typeof raw2 !== 'object' || !raw2.content) {
        throw new Error('shell::fs::read returned unexpected shape');
      }

      const { ChannelReader } = await import('iii-sdk');
      const ref: StreamChannelRef = {
        channel_id: raw2.content.channel_id,
        access_key: raw2.content.access_key,
        direction: 'read',
      };
      const reader = new ChannelReader(engineWsBase, ref);
      try {
        return await new Promise<Record<string, unknown>>((resolve, reject) => {
          const chunks: Buffer[] = [];
          let total = 0;
          let resolved = false;
          const finish = () => {
            if (resolved) return;
            resolved = true;
            try {
              reader.close();
            } catch (err) {
              logger.debug('reader.close failed', { err: String(err) });
            }
            const concat = Buffer.concat(chunks, Math.min(total, max_bytes));
            resolve(buildInlineEnvelope(concat, raw2.size));
          };
          reader.stream.on('data', (chunk: Buffer) => {
            if (resolved) return;
            const remaining = max_bytes - total;
            if (remaining <= 0) {
              finish();
              return;
            }
            const slice = chunk.length <= remaining ? chunk : chunk.subarray(0, remaining);
            chunks.push(slice);
            total += slice.length;
            if (total >= max_bytes) finish();
          });
          reader.stream.on('end', finish);
          reader.stream.on('close', finish);
          reader.stream.on('error', (err: unknown) => {
            if (resolved) return;
            resolved = true;
            try {
              reader.close();
            } catch {
              // ignore
            }
            reject(err instanceof Error ? err : new Error(String(err)));
          });
        });
      } catch (err) {
        throw new Error(`channel drain failed: ${String(err)}`);
      }
    },
    {
      description:
        'Read a host file via shell::fs::read, drain its channel, and return a {content:[{text}], details:{size, truncated, bytes_read}} envelope (max 256 KiB inline by default).',
    },
  );
}
